File: pset6/sentimental-credit/credit.js
const readline = require('readline');

// Create interface
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Prompt for card number
function promptNumber() {
    rl.question('Number: ', input => {
        if (!/^\d+$/.test(input)) {
            // not all numbers -> ask again
            promptNumber();
        } else {
            validateCard(input);
            rl.close();
        }
    });
}

// Luhn' algorithm + card check
function validateCard(numStr) {
    const digits = numStr.split('').map(Number);
    let sum1 = 0; // sum of odds
    let sum2 = 0; // sum of evens after doubling

    // Backwards iterate
    for (let i = digits.length - 1; i >= 0; i--) {
        const d = digits[i];
        const posFromRight = digits.length - i;
        if (posFromRight % 2 === 1) {
            // Odds -> add
            sum1 += d;

        } else {
            // Even -> double, then add
            const product = d * 2;
            sum2 += Math.floor(product / 10) + (product % 10);
        }
    }

    const total = sum1 + sum2;
    // If isnt 0 = Invalid
    if (total % 10 !== 0) {
        console.log('INVALID');
        return;
    }

    // Determine card type
    const length = digits.length;
    const firstOne = Number(numStr[0]);
    const firstTwo = Number(numStr.slice(0, 2));

    if (firstOne === 4 && (length === 13 || length === 16)) {
        console.log('VISA');
    } else if (firstTwo >= 51 && firstTwo <= 55 && length === 16) {
        console.log('MASTERCARD');
    } else if ((firstTwo === 34 || firstTwo === 37) && length === 15) {
        console.log('AMEX');
    } else {
        console.log('INVALID');
    }
}

// Start
promptNumber();

File: pset6/sentimental-credit/credit.py
def main():
    # Prompt for card number
    while True:
        num_str = input("Number: ").strip()
        if num_str.isdigit():
            break

    # Convert to list
    digits = [int(ch) for ch in num_str]

    # Luhn's algorithm
    # Sum of digits not multiplied
    sum1 = sum(digits[-1::-2])
    # Sum of digits after doubling every other digit
    sum2 = 0
    for d in digits[-2::-2]:
        product = d * 2
        sum2 += product // 10 + product % 10

    total = sum1 + sum2

    # If total's last digit isn't 0, not valid
    if total % 10 != 0:
        print("INVALID")
        return

    # Find card type
    length = len(digits)
    first_two = int(num_str[:2])
    first_one = int(num_str[0])

    if first_one == 4 and length in (13, 16):
        print("VISA")
    elif 51 <= first_two <= 55 and length == 16:
        print("MASTERCARD")
    elif first_two in (34, 37) and length == 15:
        print("AMEX")
    else:
        print("INVALID")


# Start
if __name__ == "__main__":
    main()
